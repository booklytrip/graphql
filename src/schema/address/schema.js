const Address = `
    # An address object
    type Address {
        # An address street
        street: String!
        
        # The postal code
        postcode: String!
        
        # The country code
        countryCode: String!
        
        # The city name
        city: String!
    }
`;

const AddressInput = `
    # An input to mutate address object
    input AddressInput {
        # An address street
        street: String!
        
        # The postal code
        postcode: String!
        
        # The country code
        countryCode: String!
        
        # The city name
        city: String!
    }
`;

export default () => [Address, AddressInput];
