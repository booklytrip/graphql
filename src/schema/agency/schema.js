const Agency = `
    type Agency implements Node {
        # The unqiue agency ID
        id: ID!
        
        # The agency name
        name: String!
        
        # An administartor of the agency
        administrator: User!
        
        # Get list of bookings across all projects of the agency
        bookings: [Booking]

        # Get list of all projects of the agency
        projects: [Project]

        # Get list of credit cards of the agency
        creditCards: [CreditCard]

        # Search queries logs of this agency
        searchQueries (
            first: Int!
            after: String
        ): SearchQueriesConnection!

        # Date of agency registration
        createdAt: String!
    }
`;

const CreateAgencyInput = `
    # An input to create a new agency
    input CreateAgencyInput {
        # The agency name (or company name)
        name: String!
        
        # The agency administrator
        administrator: CreateUserInput!
    }
`;

const CreateAgencyPayload = `
    # Response of new agency creation
    type CreateAgencyPayload {
        # A new agency entity
        agency: Agency!
        
        # A new agency administrator
        administrator: CreateUserPayload!
    }
`;

export default () => [Agency, CreateAgencyInput, CreateAgencyPayload];
