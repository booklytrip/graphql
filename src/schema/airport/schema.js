const AirportSearchResult = `
    # The airports search result
    type AirportSearchResult {
        # The result score
        _score: Float!

        # The airport object
        airport: Airport!
    }
`;

const Airport = `
    # Represents an airport object
    type Airport implements Node {
        # Represents unique identifier
        id: ID!

        # Name of airport. Mau or may not container the City name.
        name(
            # Language that the name should be presented in
            lang: String
        ): String!

        # Airport code (iata or icao, in that order first available).
        code: String!

        # Country where airport is located.
        country: Country!

        # Main city served by airport.
        city: City!

        # 3-letter FAA code, for airports located in Country "United States of America".
        # 3-letter IATA code, for all other airports.
        iata: String

        # 4-letter ICAO code.
        icao: String

        # Timezone that airport operates in.
        timezone: String
    }
`;

export default () => [Airport, AirportSearchResult];
