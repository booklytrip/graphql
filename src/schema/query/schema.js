export default `
    type Query {
        # Get currently authenticated user
        viewer: User
        
        # Search airports that matches specified term
        searchAirports(term: String! limit: Int): [AirportSearchResult]

        # Get information of airport with specified IATA code
        airport(iata: String!): Airport

        # Get single carried details
        carrier(code: String!): Carrier

        # Get list of available suppliers
        suppliers: [Supplier]

        # Get list of countries
        countries: [Country]

        # Get specific project information
        project(id: ID!): Project

        # Get specific booking information
        booking(id: ID!): Booking
    }
`;
