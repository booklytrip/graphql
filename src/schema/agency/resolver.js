import paginate from '../../lib/pagination';
import { map } from 'lodash';

export const Agency = {
    /**
     * Return list of all bookings of the aggency
     */
    bookings: async (root, args, context) => {
        const { ProjectModel, BookingModel } = context;

        // Get list of projects of the agency
        const projects = await ProjectModel.find({
            agency: root.id,
        });

        const bookings = await BookingModel.find({
            project: { $in: map(projects, project => project.id) },
        }).sort({ createdAt: -1 });

        return bookings.map(booking => ({
            ...booking.toObject(),
            totalPrice: booking.totalPrice,
        }));
    },

    /**
     * Return list of all projects of the agency
     */
    projects: (root, args, context) => {
        return context.ProjectModel.find({ agency: root.id });
    },

    /**
     * Return list of all credit cards of the agency
     */
    creditCards: (root, args, context) => {
        return context.CreditCardModel.find({ agency: root.id });
    },

    /**
     * Return list of search queries of the agency
     */
    searchQueries: async (root, args, context) => {
        const { first, after } = args;

        return paginate({
            model: context.SearchQueryModel,
            first,
            after,
            sort: { createdAt: -1 },
        });
    },
};
