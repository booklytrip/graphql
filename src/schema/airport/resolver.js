/**
 * @flow
 */

import { isEmpty, get, property } from 'lodash';

export const Airport = {
    id: property('_id'),
    code: property('iata'),

    /**
     * Return airport name in specified language.The original 
     * name will be used by default.
     */
    name: (root, args, context) => {
        if (args.lang && !isEmpty(get(root, ['translations', args.lang]))) {
            return root.translations[args.lang];
        }

        return root.name;
    },

    /**
     * Return country object where the airports operates
     */
    country: (root, args, context) => {
        return context.CountryModel.findByCca2WithCache(root.countryCode);
    },

    /**
     * Return city object where the airport operates
     */
    city: (root, args, context) => {
        return context.CityModel.findByCityCodeWithCache(root.cityCode);
    },
};
