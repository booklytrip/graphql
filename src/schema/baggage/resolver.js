import { castArray } from 'lodash';

export const Baggage = {
    cabin: baggage => {
        if (baggage.cabin) {
            return castArray(baggage.cabin);
        }
    },
    checked: baggage => {
        if (baggage.checked) {
            return castArray(baggage.checked);
        }
    },
};

export const Bag = {
    price: bag => ({
        amount: bag.price,
        currency: bag.currency,
    }),
    weight: bag => bag.weight,
};
