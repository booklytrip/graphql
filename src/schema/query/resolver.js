/**
 * @flow weak
 */

import { map } from 'lodash';

import { NotAuthenticated, NotExists } from '../../errors';

import type {
    ID,
    Context,
    FlightIdentity,
    SearchFlightsQuery,
    SearchDestionationsQuery,
} from '../../types';

export const Query = {
    /**
     * Returns current user data if user is authenticated OR
     * throws and error if not.
     */
    viewer(root, args, context) {
        if (!context.user) {
            throw new NotAuthenticated();
        }

        return context.user;
    },

    searchAirports(root, args, context: Context) {
        const { term, limit } = args;
        return context.AirportModel.search(term, limit);
    },

    airport(root, { iata }: { iata: string }, context: Context) {
        return context.AirportModel.findOne({ iata });
    },

    carrier(root, { code }: { code: string }, context: Context) {
        return context.CarrierModel.findOne({ code });
    },

    async booking(root, args, context: Context) {
        const booking = await context.BookingModel.findById(args.id);

        if (!booking) {
            throw new Error(`Booking with ID ${args.id} does not exists`);
        }

        return {
            ...booking.toObject(),
            totalPrice: booking.totalPrice,
            createdAt: booking.createdAt.toUTCString(),
        };
    },

    /**
     * Return single project identified by ID
     */
    async project(root, args, context: Context) {
        const { id } = args;

        const project = await context.ProjectModel.findById(id);
        if (!project) {
            throw new NotExists({ data: { type: 'id', id } });
        }

        return project.toObject();
    },

    /**
     * Return list of avaiable suppliers
     */
    suppliers(root, args, context: Context) {
        return context.Suppliers.find();
    },

    /**
     * Return list of countires
     */
    countries(root, args, context: Context) {
        return context.CountryModel.find().sort({ 'name.common': 1 });
    },
};
