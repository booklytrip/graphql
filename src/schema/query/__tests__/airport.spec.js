import { graphql } from 'graphql';
import { makeExecutableSchema } from 'graphql-tools';
import * as schema from '../../schema';
import * as resolvers from '../../resolver';

describe('airport', async () => {
    const executableSchema = makeExecutableSchema({
        typeDefs: Object.values(schema),
        resolvers,
    });

    it('should return an airport', async () => {
        const query = `
            query Q {
                airport(iata: "RIX") {
                    name
                }
            }
        `;

        const AirportModel = {
            findOne: () => ({
                iata: 'RIX',
                name: 'Riga',
            }),
        };

        const rootValue = {};
        const context = { AirportModel };

        const result = await graphql(
            executableSchema,
            query,
            rootValue,
            context,
        );

        expect(result.data.airport).toMatchSnapshot();
    });
});
