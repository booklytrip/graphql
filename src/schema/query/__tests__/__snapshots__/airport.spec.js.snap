// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`airport should return an airport 1`] = `
Object {
  "name": "Riga",
}
`;
